src/psyclone/psyir/nodes/psy_data_node.py

```diff
@@ -0,0 +1,536 @@
+        :type options["pre-var-list"]: list of str
+        :param options["post-var-list"]: a list of variables to be extracted \
+            after the last child.
+        :type options["post-var-list"]: list of str
+        :param str options["pre-var-postfix"]: an optional postfix that will \
+            be added to each variable name in the pre-var-list.
+        :param str options["post-var-postfix"]: an optional postfix that will \
+            be added to each variable name in the post-var-list.
+
+        '''
+        # TODO: #415 Support different classes of PSyData calls.
+        module_name = self._module_name
+        if module_name is None:
+            # The user has not supplied a module (location) name so
+            # return the psy-layer module name as this will be unique
+            # for each PSyclone algorithm file.
+            module_name = self.root.invoke.invokes.psy.name
+
+        region_name = self._region_name
+        if region_name is None:
+            # The user has not supplied a region name (to identify
+            # this particular invoke region). Use the invoke name as a
+            # starting point.
+            region_name = self.root.invoke.name
+            from psyclone.psyGen import Kern
+            kerns = self.walk(Kern)
+            if len(kerns) == 1:
+                # This PSyData region only has one kernel within it,
+                # so append the kernel name.
+                region_name += ":{0}".format(kerns[0].name)
+            # Add a region index to ensure uniqueness when there are
+            # multiple regions in an invoke.
+            psy_data_nodes = self.root.walk(PSyDataNode)
+            idx = psy_data_nodes.index(self)
+            region_name += ":r{0}".format(idx)
+
+        if not options:
+            options = {}
+
+        pre_variable_list = options.get("pre-var-list", [])
+        post_variable_list = options.get("post-var-list", [])
+        pre_suffix = options.get("pre-var-postfix", "")
+        post_suffix = options.get("post-var-postfix", "")
+
+        # Note that adding a use statement makes sure it is only
+        # added once, so we don't need to test this here!
+        use = UseGen(parent, self.fortran_module, only=True,
+                     funcnames=PSyDataNode.symbols)
+        parent.add(use)
+        var_decl = TypeDeclGen(parent, datatype="PSyDataType",
+                               entity_decls=[self._var_name],
+                               save=True)
+        parent.add(var_decl)
+
+        self._add_call("PreStart", parent,
+                       ["\"{0}\"".format(module_name),
+                        "\"{0}\"".format(region_name),
+                        len(pre_variable_list),
+                        len(post_variable_list)])
+        self.set_region_identifier(module_name, region_name)
+        has_var = pre_variable_list or post_variable_list
+
+        # All variable names get a postfix of "-post" or "-pr".
```

Comment:
This confuses me a bit as you say, a few lines down, that variable names can't contain a '-'?

Reply:
I've completely rewritten this. It was a good idea to avoid name clashes (user code could not contain a variable "a-post", so if I used "a-post" as NetCDF name, I could be certain that this name would not clash. But I need a proper symbol-table/namespace manager sooner or later anway, so I removed all of this.